### PR TITLE
Start Ydoc with the language server

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

- related #7954

Changelog:
- update: Ydoc starts with the language server on the `localhost:1234` by default. The hostname and ports can be configured by setting environment variables `LANGUAGE_SERVER_YDOC_HOSTNAME` and `LANGUAGE_SERVER_YDOC_PORT`
- update: by default `npm dev run` uses the node Ydoc server. You can control it with `POLYGLOT_YDOC_SERVER` env variable. For example,
  ```
  env POLYGLOT_YDOC_SERVER='true' npm --workspace=enso-gui2 run dev
  ```
  To connect to the Ydoc server running on the 1234 port (the one started with the language server)
  ⠀
  ```
  env POLYGLOT_YDOC_SERVER='ws://127.0.0.1:1235' npm --workspace=enso-gui2 run dev
  ```
  To connect to the provided URL. Can be useful for debugging when you start a separate Ydoc process.
- update: run `npm install` before the engine build. It is required to create the Ydoc JS bundle.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: build.sbt
Comment (Akirathan) [inline build.sbt]: At the end, the `assembly` task along with all its setting (`assembly/assemblyExcludedJars`, etc.) should not be needed. To package `ydoc-server` and run it, it should be enough to do `ydoc-server/packageBin` which will just compile classes inside that project and package it in a JAR. So at the end `ydoc-server.jar` will have just classes and resources from `ydoc-server` and nothing else.

File: build.sbt
Comment (Akirathan) [inline build.sbt:2387]: This ensures that there are no helidon classes in `runner.jar`. This means that all the helidon classes should be present in helidon modules inside the `component` directory. That is, all the helidon classes should be present on the module boot layer. Thus, modify also the `CLASS_DELEGATION_PATTERNS` in https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/ClassLoaderConstants.java#L15 to contain helidon classes as well. If you add the `org.helidon` prefix there, our boot class loader ensures that all the classes with `org.helidon` prefix will be loaded from the module boot layer (i.e. from JARs in `component` directory)

Comment (Akirathan) [reply build.sbt:2387]: By the way, I believe that if you add `org.helidon` to the `CLASS_DELEGATION_PATTERNS` (https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/ClassLoaderConstants.java#L15), it won't be necessary to remove helidon classes from `runner.jar` anymore.

File: engine/runtime-fat-jar/src/main/java/module-info.java
Comment (Akirathan) [inline engine/runtime-fat-jar/src/main/java/module-info.java:13]: Good solution for now, if everything works with that. But at the end, we must ensure that `runtime.jar` does not require helidon modules. Let's keep that in mind and create a separate ticket for that before merging.

File: engine/runtime/src/main/java/org/enso/ClassLoaderConstants.java
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/ClassLoaderConstants.java:16]: Good. This ensures that our boot class loader deleagates loading of all `io.helidon.*` classes to the system class loader, which picks the classes from JARs inside the `component` directory.

File: build.sbt
Comment (Akirathan) [reply build.sbt:2387]: I have just [noticed](https://github.com/enso-org/enso/pull/9862/files#r1598079038) that the change is already there. Sorry for the fuzz.

File: .github/workflows/engine-nightly.yml
Comment (4e6) [inline .github/workflows/engine-nightly.yml]: Python is required to run the `npm install`

File: .github/workflows/engine-nightly.yml
Comment (4e6) [inline .github/workflows/engine-nightly.yml]: `npm` is required to build the Ydoc JS bundle

**Review by somebody1234 — APPROVED** [review]
approving (minimal) changes to Dashboard code in [app/ide-desktop/lib/dashboard/src/layouts/Editor.tsx](https://github.com/enso-org/enso/pull/9862/files?w=1#diff-6055d4081368316e236105656f085ac4388bcbee81484e431c862c10fcb489a4).

File: build.sbt
Comment (hubertp) [inline build.sbt]: nit
```suggestion
  "jakarta.inject"             % "jakarta.inject-api"          % jakartaVersion
```
and define
```
val jakartaVersion = "2.0.1"
```

File: build.sbt
Comment (hubertp) [inline build.sbt:1009]: it needs **all** of those? :exploding_head:

**Review by hubertp — APPROVED** [review]
Generally the changes look fine. I'm a bit concerned that in a number of sbt projects we just add `helidon`. I doubt we need all of those dependencies in every case so maybe it makes sense to reduce that list?

File: build.sbt
Comment (4e6) [reply build.sbt:1009]: It is a way to put the classes on the module path. Although they are filtered out from the assembled JAR

File: .github/workflows/engine-nightly.yml
Comment (JaroslavTulach) [reply .github/workflows/engine-nightly.yml]: Why? Are we installing some special Python requiring `npm` module?

File: app/gui2/src/stores/project/index.ts
Comment (JaroslavTulach) [inline app/gui2/src/stores/project/index.ts]: Really `'undefined'` as a string? Don't you want to `if (!engine.ydocUrl)`? Or `typeof engine.ydocUrl === 'undefined'`?

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala]: No special option to start `Ydoc`? Do we always start it? Shouldn't there be a `--ydoc-server-port`?

File: lib/java/ydoc-server/src/main/java/module-info.java
Comment (JaroslavTulach) [inline lib/java/ydoc-server/src/main/java/module-info.java]: Why are we exporting this? Because JavaScript makes calls to those classes? If so, we should rather use `opens` instead of `exports` - as we only want to allow _reflective access_, not direct linkage.

File: project/Ydoc.scala
Comment (JaroslavTulach) [inline project/Ydoc.scala:18]: Do we still need to invoke `npm` externally when `sbt` calls it?

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java
Comment (JaroslavTulach) [inline lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java]: Better to give it a name. Now there is going to be a lot of threads and this one should be "recognizable".

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java
Comment (JaroslavTulach) [inline lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java]: Please rather store the `.js` file next to this class in the `org.enso.ydoc` package.

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
Great work! Please make sure an issue to remove `io.helidon` from `runtime.jar`. Please address some of my smaller comments.

File: tools/legal-review/engine/io.helidon.common.helidon-common-4.0.8/copyright-ignore
Comment (radeusgd) [inline tools/legal-review/engine/io.helidon.common.helidon-common-4.0.8/copyright-ignore]: We should probably not be ignoring almost all of these - I guess at least to cover the whole timespan these are representing - e.g. 2017-2024.

I think it's safest to do "when in doubt, prefer to keep".

Comment (4e6) [reply tools/legal-review/engine/io.helidon.common.helidon-common-4.0.8/copyright-ignore]: updated :+1:

File: build.sbt
Comment (4e6) [reply build.sbt]: updated :+1:

File: app/gui2/src/stores/project/index.ts
Comment (4e6) [reply app/gui2/src/stores/project/index.ts]: Yes, it's a special value because config values can only be strings.

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala]: Yes, it is always started. I agree, adding a config parameter for a port may be useful.

File: project/Ydoc.scala
Comment (4e6) [reply project/Ydoc.scala:18]: The JS bundle is required in ydoc-server tests, and when building the distribution. We need to run `npm` command to build it.

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java
Comment (4e6) [reply lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java]: okay

File: project/Ydoc.scala
Comment (JaroslavTulach) [reply project/Ydoc.scala:18]: The PR description says:

> update: run npm install before the engine build. It is required to create the Ydoc JS bundle.

Do we still need to invoke `npm install` manually & externallly when `sbt` calls it too?

File: app/gui2/src/stores/project/index.ts
Comment (Frizi) [reply app/gui2/src/stores/project/index.ts]: The config values can also be left missing, making them actually undefined here. We still should support that case.

I'm fine with adding another case here with a comment why it is necessary, but ideally i'd not modify this condition. Stringified `undefined` is often a trap that's hard to spot during debugging. In any case, the previous conditions should still be preserved.

```suggestion
    if (engine.ydocUrl == null || engine.ydocUrl === '' || engine.ydocUrl === 'undefined') {
```

File: app/gui2/vite.config.ts
Comment (Frizi) [inline app/gui2/vite.config.ts:61]: I guess this change is one of the causes for such a weird config check. The `undefined` value should not be stringified here.
```suggestion
    YDOC_SERVER_URL: POLYGLOT_YDOC_SERVER ? JSON.stringify(POLYGLOT_YDOC_SERVER) : undefined,
```

File: app/ide-desktop/lib/dashboard/src/layouts/Editor.tsx
Comment (Frizi) [inline app/ide-desktop/lib/dashboard/src/layouts/Editor.tsx]: Again unexpected stringified `undefined` that I'd like to get rid of.

**Review by Frizi — CHANGES_REQUESTED** [review]

File: app/ide-desktop/lib/dashboard/src/layouts/Editor.tsx
Comment (4e6) [reply app/ide-desktop/lib/dashboard/src/layouts/Editor.tsx]: This special value can be confusing, I agree. I refactored the `ydocUrl` logic

File: app/gui2/vite.config.ts
Comment (4e6) [reply app/gui2/vite.config.ts:61]: `JSON.stringify(undefined) === undefined` anyway. I was using this special `'undefined'` string because the config only allows having string values. I refactored the logic

File: project/Ydoc.scala
Comment (4e6) [reply project/Ydoc.scala:18]: I moved `npm install` step from CI setup to SBT build :+1:

**Review by JaroslavTulach — APPROVED** [review]
It is great to see this to be green. Let's integrate and let's unblock the cloud integration.

**Comment by somebody1234** [comment]
hey, just wondering - what's the roadmap for this stuff? will this be theoretically ready for integration with Cloud after this PR? or are there more steps that are needed?

**Comment by 4e6** [comment]
This should be enough to start integration in the cloud

**Comment by 4e6** [comment]
@Frizi waiting for your approval

**Comment by PabloBuchu** [comment]
| update: Ydoc starts with the language server on the localhost:1234 by default. The hostname and ports can be configured by setting environment variables LANGUAGE_SERVER_YDOC_HOSTNAME and LANGUAGE_SERVER_YDOC_PORT

Can you `EXPOSE 1234` port in [Dockerfile](https://github.com/enso-org/enso/blob/495eed45ba31c64bbda10cce7c7ed4f9af45f911/tools/ci/docker/Dockerfile#L4)?

The port should be fixed there and both `LANGUAGE_SERVER_YDOC_HOSTNAME` `LANGUAGE_SERVER_YDOC_PORT` overridden (same as `RPC_PORT` and `DATA_PORT`)

**Comment by 4e6** [comment]
Sure, I'll update the Docerfile in the follow-up PR. This one has already too many stuff in it

**Review by Frizi — APPROVED** [review]

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java
Comment (AdRiley) [inline lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java:20]: Is it intentional that this Semaphore is never released?

**Review by AdRiley — APPROVED** [review]

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java
Comment (4e6) [reply lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java:20]: yes, we are waiting for any interruption signal here. Otherwise, the main process ends.

**Comment by radeusgd** [comment]
I realise that once this was merged, I can no longer compile the engine through sbt in IntelliJ, because I use the ([recommended](https://github.com/enso-org/enso/blob/develop/docs/CONTRIBUTING.md#system-requirements)) version manager for Node: `fnm`. Unfortunately I was never able to set up IntelliJ to pick-up the environment changes that `fnm` generates to be able to 'see' `node` from an `sbt` shell in IntelliJ.

For now I just need to switch to using sbt from within PowerShell. This is okay, but a bit annoying as IntelliJ will start the sbt shell anyway to re-import the project from time to time.

Has anyone had a chance to set up IntelliJ to pick up the Node paths from `fnm`?

**Comment by Akirathan** [comment]
> I realise that once this was merged, I can no longer compile the engine through sbt in IntelliJ, because I use the ([recommended](https://github.com/enso-org/enso/blob/develop/docs/CONTRIBUTING.md#system-requirements)) version manager for Node: `fnm`. Unfortunately I was never able to set up IntelliJ to pick-up the environment changes that `fnm` generates to be able to 'see' `node` from an `sbt` shell in IntelliJ.
> 
> For now I just need to switch to using sbt from within PowerShell. This is okay, but a bit annoying as IntelliJ will start the sbt shell anyway to re-import the project from time to time.
> 
> Has anyone had a chance to set up IntelliJ to pick up the Node paths from `fnm`?

@radeusgd My IntelliJ's builtin sbt shell stopped working approx 9 months ago. I have switched to a separate `sbt` process in the terminal entirely. It is interesting that you have even made it so far.

**Pull request merged**
